### DISCUSSION TITLE
Consider environment for connection

### DISCUSSION THREAD

**Discussion by ASMailer** [Micke-K/IntuneManagement#105] [General]
Hey Mike,

we want to use the tool in an different graph environment and configured in the settings "show azure AD login menu".
Now we prompted and can select a different environment, and sign-in works perfect. We also added a custom application ID for the graph access. What I'm now searching, how to change the msgraph environment (similar like in MgGraph). Is there a possibility to set a different graph environment in one of the init scripts ?

thx
Andi

**Comment by Micke-K** [comment]
Hello Andi?

Is this for the Silent Batch jobs? Or just logging in to via an app?

I don't think either of them is possible at the moment. I have to revisit these scenarios to understand them again. Logging in via an app might be possible by specifying the Authority setting but that is not available in the batch login.

I know that MSAL has support for login in to additional environments via the public authority but I haven't looked at that and I have no way of testing this myself.

Cheers!

**Comment by ASMailer** [comment]
Hi Mike,

my use case is about connecting graph e.g. china environment. The behavior is depending on the used module.

In case using MSGraph module I need to update the MSGraphEnvironment with corresponding application ID (graphAPI), AuthUrl, GraphBaseUrl, GraphResourceId.

... here an example ...
Update-MSGraphEnvironment -AppId "xxxxxxxxxx-xxxx-xxxx-xxxxxxxxxxx" -AuthUrl 'https://login.chinacloudapi.cn/common' -GraphBaseUrl 'https://microsoftgraph.chinacloudapi.cn' -GraphResourceId 'https://microsoftgraph.chinacloudapi.cn' -SchemaVersion 'v1.0'

After update the environment I´m able to run connect-msgraph and run cmdlets, etc...

In case of using the MGGraph there is the possibility to define china environment in the command line, for example ...
Connect-MgGraph -Environment China -ClientId xxxxxxxxxx-xxxx-xxxx-xxxxxxxxxxx -TenantId xxxxxxxxxx-xxxx-xxxx-xxxxxxxxxxx

I changed in the settings the "show azure AD login menu" option and added in the application ID property our graph app ID.
By doing so I get the right prompt and was able to authenticate against the china environment without an error of the application ID.
That was the reason why I was thinking that also the custom application ID was considered.

But I received a error message that "graph.microsoft.com" was not available.

My idea now is to find the position, where the connect is executed, and adjust the MSGraphEnvironment as posted above.

Greetings
Andi

**Comment by ASMailer** [comment]
Hi Mike 

I think the problem is fixed, you have already considered all (azure environments) of this, and the reason why it hasn´t worked was because of some little typos ... :-)

MSALAuthentication.psm1
![image](https://user-images.githubusercontent.com/79043883/207629178-ee3b457e-4b99-4748-b985-b9ff72040b8e.png)
... the "**https://**" needs to be removed ...

![image](https://user-images.githubusercontent.com/79043883/207629432-6ca5885e-61dc-4bc8-90ff-f2ece9cef9a4.png)
... the **"r"** was missing ...

It was then sufficient to adjust the following settings, so that it worked then ...
![image](https://user-images.githubusercontent.com/79043883/207631011-3cf37cf0-c176-4564-a6c0-f8ca4120d014.png)

... really well prepared the tool, awesome!

should i create a issue, or fits this info for you ?

Greetings
Andi

**Comment by Micke-K** [comment]
Hello Andi,

Amazing find! Thank you so much. 

And I am sorry about that. One of the things I have no way of testing. I really appreciate the help!

I'll convert this into an issue and fix this in the next release. 

Cheers!
